File: pytet_v0.4/ctetris.py
from tetris import *

class CTetris(Tetris):
    setOfColorBlockObjects = 0


    def init(setOfBlockArrays):
        Tetris.nBlockTypes = len(setOfBlockArrays)
        Tetris.nBlockDegrees = len(setOfBlockArrays[0])
        Tetris.setOfBlockObjects = [[0] * Tetris.nBlockDegrees for _ in range(Tetris.nBlockTypes)]
        arrayBlk_maxSize = 0
        for i in range(Tetris.nBlockTypes):
            if arrayBlk_maxSize <= len(setOfBlockArrays[i][0]):
                arrayBlk_maxSize = len(setOfBlockArrays[i][0])
        Tetris.iScreenDw = arrayBlk_maxSize     # larget enough to cover the largest block

        for i in range(Tetris.nBlockTypes):
            for j in range(Tetris.nBlockDegrees):
                Tetris.setOfBlockObjects[i][j] = Matrix(setOfBlockArrays[i][j])

        CTetris.setOfColorBlockObjects = [[0] * Tetris.nBlockDegrees for _ in range(Tetris.nBlockTypes)]

        for i in range(Tetris.nBlockTypes):
            for j in range(Tetris.nBlockDegrees):
                CTetris.setOfColorBlockObjects[i][j] = Matrix(setOfBlockArrays[i][j])
                CTetris.setOfColorBlockObjects[i][j].mulc(i+1)
                print(CTetris.setOfColorBlockObjects[i][j])
        return 

    def __init__(self, iScreenDy, iScreenDx):
        super().__init__(iScreenDy, iScreenDx)
        arrayColorScreen = self.createArrayScreen()
        self.iCScreen = Matrix(arrayColorScreen)
        self.oCScreen = Matrix(self.iScreen)
        return
        
    def accept(self, key):
        self.state = TetrisState.Running

        if key >= '0' and key <= '6':
            if self.justStarted == False:
                self.deleteFullLines()
            self.iScreen = Matrix(self.oScreen)
            self.iCScreen = Matrix(self.oCScreen)
            self.idxBlockType = int(key)
            self.idxBlockDegree = 0
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            self.currColorBlk = self.setOfColorBlockObjects[self.idxBlockType][self.idxBlockDegree]
            self.top = 0
            self.left = Tetris.iScreenDw + self.iScreenDx//2 - self.currBlk.get_dx()//2
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            self.tempColorBlk = self.iCScreen.clip(self.top, self.left, self.top+self.currColorBlk.get_dy(), self.left+self.currColorBlk.get_dx())
            self.tempBlk = self.tempBlk + self.currBlk
            self.tempColorBlk = self.tempColorBlk + self.currColorBlk
            self.justStarted = False
            print()

            if self.tempBlk.anyGreaterThan(1):
                self.state = TetrisState.Finished
            self.oScreen = Matrix(self.iScreen)
            self.oScreen.paste(self.tempBlk, self.top, self.left)
            self.oCScreen = Matrix(self.iCScreen)
            self.oCScreen.paste(self.tempColorBlk, self.top, self.left)
            return self.state
        elif key == 'q':
            pass
        elif key == 'a': # move left
            self.left -= 1
        elif key == 'd': # move right
            self.left += 1
        elif key == 's': # move down
            self.top += 1
        elif key == 'w': # rotate the block clockwise
            self.idxBlockDegree = (self.idxBlockDegree + 1) % Tetris.nBlockDegrees
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            self.currColorBlk = self.setOfColorBlockObjects[self.idxBlockType][self.idxBlockDegree]
        elif key == ' ': # drop the block
            while not self.tempBlk.anyGreaterThan(1):
                    self.top += 1
                    self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
                    self.tempBlk = self.tempBlk + self.currBlk
                    self.tempColorBlk = self.iCScreen.clip(self.top, self.left, self.top+self.currColorBlk.get_dy(), self.left+self.currColorBlk.get_dx())
                    self.tempColorBlk = self.tempColorBlk + self.currColorBlk
        else:
            print('Wrong key!!!')
            
        self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
        self.tempBlk = self.tempBlk + self.currBlk
        self.tempColorBlk = self.iCScreen.clip(self.top, self.left, self.top+self.currColorBlk.get_dy(), self.left+self.currColorBlk.get_dx())
        self.tempColorBlk = self.tempColorBlk + self.currColorBlk

        if self.tempBlk.anyGreaterThan(1):   ## 벽 충돌시 undo 수행
            if key == 'a': # undo: move right
                self.left += 1
            elif key == 'd': # undo: move left
                self.left -= 1
            elif key == 's': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            elif key == 'w': # undo: rotate the block counter-clockwise
                self.idxBlockDegree = (self.idxBlockDegree - 1) % Tetris.nBlockDegrees
                self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
                self.currColorBlk = self.setOfColorBlockObjects[self.idxBlockType][self.idxBlockDegree]
            elif key == ' ': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            self.tempBlk = self.tempBlk + self.currBlk
            self.tempColorBlk = self.iCScreen.clip(self.top, self.left, self.top+self.currColorBlk.get_dy(), self.left+self.currColorBlk.get_dx())
            self.tempColorBlk = self.tempColorBlk + self.currColorBlk

        self.oScreen = Matrix(self.iScreen)
        self.oScreen.paste(self.tempBlk, self.top, self.left)
        self.oCScreen = Matrix(self.iCScreen)
        self.oCScreen.paste(self.tempColorBlk, self.top, self.left)
        return self.state

    def deleteFullLines(self):
        array = self.oScreen.get_array()
        colorArray = self.oCScreen.get_array()
        for y in range(self.iScreenDy):
            fullLine = True
            for x in range (self.iScreenDx):
                if array[y][Tetris.iScreenDw + x] != 1: 
                    fullLine = False
                    break
            if fullLine == True: 
                for row in range(y, 0, -1): 
                    array[row] = array[row-1]
                    colorArray[row] = colorArray[row-1]
                for x in range(Tetris.iScreenDw):
                    array[0][x] = 1
                    colorArray[0][x] = 1
                for x in range(self.iScreenDx):
                    array[0][Tetris.iScreenDw + x] = 0
                    colorArray[0][Tetris.iScreenDw + x] = 0
                for x in range(Tetris.iScreenDw):
                    array[0][Tetris.iScreenDw + self.iScreenDx + x] = 1
                    colorArray[0][Tetris.iScreenDw + self.iScreenDx + x] = 1
        self.iScreen = Matrix(array)
        self.iScreen.paste(self.tempBlk, self.top, self.left)
        self.iCScreen = Matrix(colorArray)
        self.iCScreen.paste(self.tempColorBlk, self.top, self.left)
        return

File: cpptet_v0.5/CTetris.cpp
#include "CTetris.h"
#include <cmath>

CTetris::CTetris(int dy, int dx) : Tetris(dy, dx){
    iScreenDy = dy;
    iScreenDx = dx;
    idxBlockDegree = 0;

    int arrayBlk_maxSize = 0;
    for(int i = 0; i < nBlockTypes; i++){
        int n = setOfBlockObjects[i][0]->get_dx();
        if(arrayBlk_maxSize < n) arrayBlk_maxSize = n;
    }
    iScreenDw = arrayBlk_maxSize;


    arrayCScreen = createArrayScreen();
    Matrix *icsmat = new Matrix(arrayCScreen,iScreenDy+iScreenDw,iScreenDx+iScreenDw*2);
    iCScreen = icsmat;
    Matrix *ocsmat = new Matrix(iCScreen);
    oCScreen = ocsmat;

    arrayScreen = createArrayScreen();
    Matrix *ismat = new Matrix(arrayScreen,iScreenDy+iScreenDw,iScreenDx+iScreenDw*2);
    iScreen = ismat;
    Matrix *osmat = new Matrix(iScreen);
    oScreen = osmat;

    justStarted = true;
}

CTetris::~CTetris(){
    cout << "delete setOfColorBlockObjects" << endl;
    while(!setOfColorBlockObjects.empty()){
        vector<Matrix*> tmp_layer = setOfColorBlockObjects.back();
        setOfColorBlockObjects.pop_back();
        while(!tmp_layer.empty()){
            Matrix* tmp = tmp_layer.back();
            tmp_layer.pop_back();
            delete tmp;
        }
    }

    cout << "delete arrayCScreen" << endl;
    delete arrayCScreen;
    cout << "delete iCScreen" << endl;
    delete iCScreen;
    cout << "delete oCScreen" << endl;
    delete oCScreen;
    cout << "delete currColorBlk" << endl;
    delete currColorBlk;
    cout << "delete tempColorBlk" << endl;
    delete tempColorBlk;
}

void CTetris::init(int *setOfCBlockArrays[], int blk_type, int blk_degree){
    nBlockTypes = blk_type;
    nBlockDegrees = blk_degree;

    for (int i = 0; i < blk_type; i++){
        vector<Matrix*> setOfOneBlock;
        int m = 0;
        while(1){
            if (*(setOfCBlockArrays[i*blk_degree]+m) == -1) break;
            m++;
        }
        int n = sqrt(m);

        for(int j = 0; j < blk_degree; j++) {
            setOfOneBlock.push_back(new Matrix(setOfCBlockArrays[i*blk_degree+j],n,n));
        }
        setOfBlockObjects.push_back(setOfOneBlock);
    }

    for(int j =0; j < nBlockTypes; j++){
        vector<Matrix*> setOfOneColorBlock;
        for(int i =0; i < nBlockDegrees; i++){
            Matrix *mat = new Matrix(setOfBlockObjects[j][i]);
            mat->mulc(j+2);
            setOfOneColorBlock.push_back(mat);
        }
        setOfColorBlockObjects.push_back(setOfOneColorBlock);
    }

}

TetrisState CTetris::accept(char key){
    state = Running;
    if (key >= '0' and key <= '6'){
        if (justStarted == false)
            deleteFullLines();
        
        idxBlockType = key - '0';
        idxBlockDegree = 0;

        Matrix *ismat = new Matrix(oScreen);
        iScreen = ismat;

        currBlk = setOfBlockObjects[idxBlockType][idxBlockDegree];
        
        Matrix *icsmat = new Matrix(oCScreen);
        iCScreen = icsmat;

        currColorBlk = setOfColorBlockObjects[idxBlockType][idxBlockDegree];

        top = 0;
        left = iScreenDw + iScreenDx/2 - currBlk->get_dx()/2;

        tempBlk = iScreen->clip(top, left, top+currBlk->get_dy(), left+currBlk->get_dx());
        tempBlk = tempBlk->add(currBlk);

        tempColorBlk = iCScreen->clip(top, left, top+currColorBlk->get_dy(), left+currColorBlk->get_dx());
        tempColorBlk = tempColorBlk->add(currColorBlk);

        justStarted = false;
        cout << endl;

        if (tempBlk->anyGreaterThan(1))
            state = Finished;

        Matrix *osmat = new Matrix(iScreen);
        oScreen = osmat;
        oScreen->paste(tempBlk, top, left);
        
        Matrix *ocsmat = new Matrix(iCScreen);
        oCScreen = ocsmat;
        oCScreen->paste(tempColorBlk, top, left);

        return state;
    }
    else if (key == 'q');

    else if (key == 'a') // move left
        left -= 1;
    else if (key == 'd') // move right
        left += 1;
    else if (key == 's') // move down
        top += 1;
    else if (key == 'w'){ // rotate the block clockwise
        idxBlockDegree = (idxBlockDegree + 1) % nBlockDegrees;
        currBlk = setOfBlockObjects[idxBlockType][idxBlockDegree];
        currColorBlk = setOfColorBlockObjects[idxBlockType][idxBlockDegree];
    }
    else if (key == ' '){ // drop the block
        while (!tempBlk->anyGreaterThan(1)){
            top += 1;
            tempBlk = iScreen->clip(top, left, top+currBlk->get_dy(), left+currBlk->get_dx());
            tempBlk = tempBlk->add(currBlk);
            tempColorBlk = iCScreen->clip(top, left, top+currColorBlk->get_dy(), left+currColorBlk->get_dx());
            tempColorBlk = tempColorBlk->add(currColorBlk);
        }
    }
    else
        cout << "Wrong key!!!\n";

    //delete tempBlk;
    tempBlk = iScreen->clip(top, left, top+currBlk->get_dy(), left+currBlk->get_dx());
    tempBlk = tempBlk->add(currBlk);    

    //delete tempColorBlk;
    tempColorBlk = iCScreen->clip(top, left, top+currColorBlk->get_dy(), left+currColorBlk->get_dx());
    tempColorBlk = tempColorBlk->add(currColorBlk);

    if (tempBlk->anyGreaterThan(1)){   // 벽 충돌시 undo 수행
        if (key == 'a') // undo: move right
            left += 1;
        else if (key == 'd') // undo: move left
            left -= 1;
        else if (key == 's'){ // undo: move up
            top -= 1;
            state = NewBlock;
        }
        else if (key == 'w'){ // undo: rotate the block counter-clockwise
            idxBlockDegree = (idxBlockDegree - 1) % nBlockDegrees;
            //delete currBlk;
            currBlk = setOfBlockObjects[idxBlockType][idxBlockDegree];
            //delete currColorBlk;
            currColorBlk = setOfColorBlockObjects[idxBlockType][idxBlockDegree];
        }
        else if (key == ' '){ // undo: move up
            top -= 1;
            state = NewBlock;
        }

        tempBlk = iScreen->clip(top, left, top+currBlk->get_dy(), left+currBlk->get_dx());
        tempBlk = tempBlk->add(currBlk);

        tempColorBlk = iCScreen->clip(top, left, top+currColorBlk->get_dy(), left+currColorBlk->get_dx());
        tempColorBlk = tempColorBlk->add(currColorBlk);
    }

    Matrix *osmat = new Matrix(iScreen);
    oScreen = osmat;
    oScreen->paste(tempBlk, top, left);
    
    Matrix *ocsmat = new Matrix(iCScreen);
    oCScreen = ocsmat;
    oCScreen->paste(tempColorBlk, top, left);

    return state;
}

int* CTetris::createArrayScreen(){
    int arrayScreenDx = iScreenDw * 2 + iScreenDx;
    int arrayScreenDy = iScreenDy + iScreenDw;
    int* arrayScreen = (int*)malloc(sizeof(int)*(arrayScreenDy*arrayScreenDx));

    for(int y = 0; y < arrayScreenDy; y++){
        for (int x = 0; x < iScreenDw; x++)
            arrayScreen[y*arrayScreenDx + x] = 1;
        for (int x = 0; x < iScreenDx; x++)
            arrayScreen[y*arrayScreenDx+iScreenDw + x] = 0;
        for (int x = 0; x < iScreenDw; x++)
            arrayScreen[y*arrayScreenDx+iScreenDw + iScreenDx + x] = 1;
    }

    for (int y = 0; y < iScreenDw; y++){
        for (int x = 0; x < arrayScreenDx; x++)
            arrayScreen[(iScreenDy + y)*arrayScreenDx + x] = 1;
    }
    return arrayScreen;
}

void CTetris::deleteFullLines(){
    int **array = oScreen->get_array();
    int **colorArray = oCScreen->get_array();
    for(int y = 0; y<iScreenDy; y++){
        bool fullLine = true;
        for (int x =0; x < iScreenDx; x++){
            if (array[y][iScreenDw + x] != 1){
                fullLine = false;
                break;
            }
        }
        if (fullLine == true){
            for (int row = y; row > 0; row--){
                array[row] = array[row-1];
                colorArray[row] = colorArray[row-1];
            }
            for (int x=0; x< iScreenDw; x++){
                array[0][x] = 1;
                colorArray[0][x] = 1;
            }
            for (int x=0; x<iScreenDx; x++){
                array[0][iScreenDw + x] = 0;
                colorArray[0][iScreenDw + x] = 0;
            }
            for (int x=0; x< iScreenDw; x++){
                array[0][iScreenDw + iScreenDx + x] = 1;
                colorArray[0][iScreenDw + iScreenDx + x] = 1;
            }
        }
    }

    iScreen = new Matrix(*array, iScreenDy, iScreenDx);
    iScreen->paste(tempBlk, top, left);

    iCScreen = new Matrix(*colorArray, iScreenDy, iScreenDx);
    iCScreen->paste(tempColorBlk, top, left);
}